### PR TITLE
- Ensure that the engine is not accessible after it has been closed.

### PR DESCRIPTION
**Pull request by baldersheim**
- Also protect against multiple close.

@bratseth PR

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java [hunk 1]
```diff
@@ -55,10 +57,11 @@
 
     private final Clock clock;
     private final String dataDir;
-    private final CairoEngine engine;
+    private final AtomicReference<CairoEngine> engine = new AtomicReference<>();
+    // TODO Is this safe, or even possible. ThreadLocal members should be static
```
Comment (bratseth) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java:61]: There's only ever one instance of this class so it doesn't matter, but having a threadlocal per instance just means you'll have one threadlocal item per thread+instance. Here the threadlocal is just to create fewer instances, there's no constraint that there must be max one per thread.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java [hunk 2]
```diff
@@ -222,7 +229,7 @@ private void ensureNodeTableIsUpdated() {
 
     private void ensureClusterTableIsUpdated() {
         try {
-            if (0 == engine.getStatus(newContext().getCairoSecurityContext(), new Path(), clusterTable.name)) {
+            if (0 == engine.get().getStatus(newContext().getCairoSecurityContext(), new Path(), clusterTable.name)) {
```
Comment (bratseth) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java:232]: Now we need to check for null here too.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java [hunk 3]
```diff
@@ -196,8 +199,12 @@ public void gc() {
 
     @Override
     public void close() {
-        if (engine != null)
-            engine.close();
+        synchronized (clusterTable.writeLock) {
+            CairoEngine myEngine = engine.getAndSet(null);
```
Comment (bratseth) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java:203]: This seems a little too complicated to me, why not just have an atomic boolean "closed"?
But we still need to check for it everywhere - do we actually have any problems with this?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java (hunk 1 above)
Comment (baldersheim) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java]: Ok, I concluded that was the case too. However I do not think that you want to use thread locals for this. The SqlCompilers needs to be closed. I assume prior to the engine too, but that you are not able to with thread locals.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java [hunk 4]
```diff
@@ -196,8 +199,12 @@ public void gc() {
 
     @Override
     public void close() {
-        if (engine != null)
-            engine.close();
+        synchronized (clusterTable.writeLock) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java]: I noticed, just removed it.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java (hunk 2 above)
Comment (baldersheim) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java:238]: Alternatively, just wait for the nullptr exception and fix the incorrect usage. Nothing should be using the engine after it been closed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java (hunk 1 above)
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java]: I was advised to do it by one of the QuestDb authors, but you are right they have a close() though. But have we seen any problems related to closing?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java (hunk 3 above)
Comment (baldersheim) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java]: We have a issue that shows up in JNI or very lowlevel(UnSafe) java code that we can not explain.
Then it is at least good to verify that we are not accessing objects after they are invalidated.

Using an AtomicBoolean is fine and add checks. This was just a suggestion. There might be better ways to do this.
I do not know the threading model here so I can not say what is the best approach.
But detecting use after close, and closing down the SqlCompilers are minimum I think.
I see that the SqlExecutionContext is Closable too, but calling close() on it seems dangerous as it seems to close a shared resource(MessageBus).

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java (hunk 1 above)
Comment (baldersheim) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java]: I have not looked in detail at the core dumps so I can not say. I guess we need to dive in.

**Comment by bratseth** [comment]
Can I work a bit on this PR now? Seems the simplest is to just be extremely conservative with how we invoke it.

**Comment by baldersheim** [comment]
Go ahead.

**Comment by bratseth** [comment]
@baldersheim please take a look at the last commit. 

This still doesn't protect against concurrent *read* and close. If we need that too it's starting to become cumbersome ...

**Comment by bratseth** [comment]
@baldersheim PTAL

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java [hunk 5]
```diff
@@ -196,8 +201,15 @@ public void gc() {
 
     @Override
     public void close() {
-        if (engine != null)
-            engine.close();
+        if (closed.get()) return;
+        closed.set(true);
```
Comment (baldersheim) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java:205]: you need to use if (closed.getAndSet(true)) return to make it atomic.
If not there is a race betwene get() and set(true).

**Review by baldersheim — COMMENTED** [review]
Apart from getAndSet you should be fine.

**Pull request merged**